pypeclub/OpenPype
# General: Validate if current process OpenPype version is requested version

## Brief description
Added periodical checks of OpenPype version.

## Description
Main change is affecting OpenPype tray application. A dialog with information about outdated OpenPype version is shown. If dialog is closed or ignore button is clicked a tray message will popup.

Ftrack event server statuser will show information if is up to date.

## Changes
- moved few functions in lib and added new for checking if version of currently running process is up to date
- tray will check each 5 minutes if is up to date with expected version
    - will show a dialog with ability to restart tray
- restart of tray will unset `OPENPYPE_VERSION` environment for new process
    - this is to trigger update if needed
- ftrack event server statuser has new category OpenPype version with info about current OpenPype version

## Testing notes:
1. Build OpenPype
2. Create a zip with higher version
3. Optional: Remove versions from appdata...
4. Set path to openpype versions to some empty directory
5. Run tray from the build
6. Move the created zip with higher version to openpype versions dir
7. Wait for 5 minutes
8. Popup should be shown
    - you can press Ignore window will close and tray message will show
    - press "Restart and Change" should trigger restart of tray and also update

### Alternative to not wait 5 minutes (code modification needed)
In build go to `~/openpype/tools/tray/pype_tray.py` on line 375 is method `_on_version_action` modify it to
```
def _on_version_action(self):
    self._version_check_timer.timeout.emit()
    ...
```
Now when you click on version action in tray menu it should trigger version validation.

## Notes
- any changes of messages are welcomed
- we could maybe change color of tray icon or add icon to version action (e.g. exclamation mark) or just change color of action text
- color of "Restart and Change" could be maybe changed to better color?
- It can happen that it is not possible to validate if current version is that which should be used. Should we tell user on start?

### Screenshot of dialog showed from Tray
![image](https://user-images.githubusercontent.com/43494761/149374353-63867d47-9367-4b90-8acd-a4d8db9ecf53.png)
